src/output/diagnosis.rs

```diff
@@ -0,0 +1,327 @@
+// (c) Meta Platforms, Inc. and affiliates.
+//
+// Use of this source code is governed by an MIT-style
+// license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+use crate::output as tv;
+use crate::spec;
+use tv::dut;
+
+/// This structure represents a Diagnosis message.
+/// ref: https://github.com/opencomputeproject/ocp-diag-core/tree/main/json_spec#diagnosis
+///
+/// Information about the source file and line number are not automatically added.
+/// Add them using the builder or the macros octptv_diagnosis_*
+///
+/// # Examples
+///
+/// ## Create a Diagnosis object with the `new` method
+///
+/// ```
+/// use ocptv::output::Diagnosis;
+/// use ocptv::output::DiagnosisType;
+///
+/// let diagnosis = Diagnosis::new("verdict", DiagnosisType::Pass);
+/// ```
+///
+/// ## Create a Diagnosis object with the `builder` method
+///
+/// ```
+/// use ocptv::output::HardwareInfo;
+/// use ocptv::output::Diagnosis;
+/// use ocptv::output::DiagnosisType;
+/// use ocptv::output::Subcomponent;
+///
+/// let diagnosis = Diagnosis::builder("verdict", DiagnosisType::Pass)
+///     .hardware_info(&HardwareInfo::builder("id", "name").build())
+///     .message("message")
+///     .subcomponent(&Subcomponent::builder("name").build())
+///     .source("file.rs", 1)
+///     .build();
+/// ```
+pub struct Diagnosis {
+    verdict: String,
+    diagnosis_type: spec::DiagnosisType,
+    message: Option<String>,
+    hardware_info: Option<tv::HardwareInfo>,
+    subcomponent: Option<tv::Subcomponent>,
+    source_location: Option<spec::SourceLocation>,
+}
+
+impl Diagnosis {
+    /// Builds a new Diagnosis object.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use ocptv::output::Diagnosis;
+    /// use ocptv::output::DiagnosisType;
+    ///
+    /// let diagnosis = Diagnosis::new("verdict", DiagnosisType::Pass);
+    /// ```
+    pub fn new(verdict: &str, diagnosis_type: spec::DiagnosisType) -> Self {
+        Diagnosis {
+            verdict: verdict.to_string(),
```

Comment:
nit; `.to_owned()` here better encapsulates the semantic that youre using; in reality, `.to_string()` comes from a blanket impl of trait Display as opposed of coming from `str` directly (which to_owned does)